src/Evolution/Systems/GrMhd/GhValenciaDivClean/BoundaryConditions/BoundaryCondition.hpp

```diff
@@ -0,0 +1,83 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <pup.h>
+
+#include "Domain/BoundaryConditions/BoundaryCondition.hpp"
+#include "Domain/BoundaryConditions/None.hpp"
+#include "Domain/BoundaryConditions/Periodic.hpp"
+#include "Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/BoundaryCondition.hpp"
+#include "Evolution/Systems/GeneralizedHarmonic/BoundaryConditions/Factory.hpp"
+#include "Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryConditions/BoundaryCondition.hpp"
+#include "Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryConditions/Factory.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \brief Boundary conditions for the combined Generalized Harmonic and
+/// Valencia GRMHD systems
+namespace grmhd::GhValenciaDivClean::BoundaryConditions {
+/// \cond
+template <typename DerivedGhCondition, typename DerivedValenciaCondition>
+class ProductOfConditions;
+/// \endcond
+
+namespace detail {
+template <typename DerivedGhCondition, typename DerivedValenciaCondition>
+using ProductOfConditionsIfConsistent = tmpl::conditional_t<
+    DerivedGhCondition::bc_type == DerivedValenciaCondition::bc_type,
+    ProductOfConditions<DerivedGhCondition, DerivedValenciaCondition>,
+    tmpl::list<>>;
+
+template <typename GhList, typename ValenciaList>
+struct AllProductConditions;
+
+template <typename GhList, typename... ValenciaConditions>
+struct AllProductConditions<GhList, tmpl::list<ValenciaConditions...>> {
+  using type = tmpl::flatten<tmpl::list<tmpl::transform<
+      GhList, tmpl::bind<ProductOfConditionsIfConsistent, tmpl::_1,
+                         tmpl::pin<ValenciaConditions>>>...>>;
+};
+}  // namespace detail
+
+/// \brief The base class for Generalized Harmonic and Valencia combined
+/// boundary conditions; all boundary conditions for this system must inherit
+/// from this base class.
+class BoundaryCondition : public domain::BoundaryConditions::BoundaryCondition {
+ public:
+  // remove the periodic and none BCs from the creatable classes of the
+  // individual systems; for the remaining conditions, include a
+  // `ProductOfConditions` for each pair with compatible `bc_type`s.
+  using creatable_classes = tmpl::push_back<
```

Comment:
I agree that this tmp logic isn't terribly elegant -- I think I prefer the current version to explicitly listing the matched pairs of types, because it is more extensible to the changes that I'll be making soon that relax the requirement of matched bc_type (so then the list would become ~n^2 rather than ~n).  Also, I do think that the `Periodic` and `None` don't have `bc_type`s in the same way, so can't be filtered as easily.

Though, I think this will become far more readable if I just define a helper meta-lambda so the filter logic doesn't need to be repeated, so I'll do that.

Reply:
That future change (leading to n^2 vs n pairs) is a pretty convincing argument in favor of the current setup, heh.

I like the idea of moving some of the repeated filter logic into a helper -- up to you if you'd like to do that now or in a followup PR :)